gcloud-java-core/src/main/java/com/google/gcloud/BaseIamPolicy.java

```diff
@@ -0,0 +1,297 @@
+    }
+
+    private Identity(Type type, String id) {
+      this.type = type;
+      this.id = id;
+    }
+
+    public Type type() {
+      return type;
+    }
+
+    /**
+     * Returns the string identifier for this identity. The id corresponds to:
+     * <ul>
+     *   <li>email address (for identities of type {@code USER}, {@code SERVICE_ACCOUNT}, and
+     *       {@code GROUP})
+     *   <li>domain (for identities of type {@code DOMAIN})
+     *   <li>null (for identities of type {@code ALL_USERS} and {@code ALL_AUTHENTICATED_USERS})
+     * </ul>
+     */
+    public String id() {
+      return id;
+    }
+
+    /**
+     * Returns a new identity representing anyone who is on the internet; with or without a Google
+     * account.
+     */
+    public static Identity allUsers() {
+      return new Identity(Type.ALL_USERS, null);
+    }
+
+    /**
+     * Returns a new identity representing anyone who is authenticated with a Google account or a
+     * service account.
+     */
+    public static Identity allAuthenticatedUsers() {
+      return new Identity(Type.ALL_AUTHENTICATED_USERS, null);
+    }
+
+    /**
+     * Returns a new user identity.
+     *
+     * @param email An email address that represents a specific Google account. For example,
+     *     alice@gmail.com or joe@example.com.
+     */
+    public static Identity user(String email) {
+      return new Identity(Type.USER, checkNotNull(email));
+    }
+
+    /**
+     * Returns a new service account identity.
+     *
+     * @param email An email address that represents a service account. For example,
+     *     my-other-app@appspot.gserviceaccount.com.
+     */
+    public static Identity serviceAccount(String email) {
+      return new Identity(Type.SERVICE_ACCOUNT, checkNotNull(email));
+    }
+
+    /**
+     * Returns a new group identity.
+     *
+     * @param email An email address that represents a Google group. For example,
+     *     admins@example.com.
+     */
+    public static Identity group(String email) {
+      return new Identity(Type.GROUP, checkNotNull(email));
+    }
+
+    /**
+     * Returns a new domain identity.
+     *
+     * @param domain A Google Apps domain name that represents all the users of that domain. For
+     *     example, google.com or example.com.
+     */
+    public static Identity domain(String domain) {
+      return new Identity(Type.DOMAIN, checkNotNull(domain));
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(id, type);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (!(obj instanceof Identity)) {
+        return false;
+      }
+      Identity other = (Identity) obj;
+      return Objects.equals(id, other.id()) && Objects.equals(type, other.type());
+    }
+  }
+
+  /**
+   * Builder for an IAM Policy.
+   */
+  protected abstract static class BaseBuilder<R, B extends BaseBuilder<R, B>> {
+
+    private final Map<R, List<Identity>> bindings = new HashMap<>();
```

Comment:
see other related comments. Maybe this should be `Map<R, Set<Identity>>` and use `LinedHashSet` for maintaining the order.